### ISSUE TITLE
DocumentSession.SaveChangesAsync() loads document from wrong tenant when applying events from another tenant to an inline projection

### ISSUE THREAD

**Issue by iamjustsam** [JasperFx/marten#2363]
This issue is related to the functionality implemented for #2298 

When opening a session in `Tenant A` and then applying events for another tenant, Marten tries to load the projected document from the tenant of the session instead of the tenant the document actually belongs to.

Unit test to point out the issue:
```cs
using System;
using System.Threading.Tasks;
using Marten;
using Marten.Events.Projections;
using Marten.Storage;
using NUnit.Framework;

namespace TestProject1;

public class Tests
{
    private IDocumentStore _store;
    private Guid _aggregateId;

    [SetUp]
    public async Task Setup()
    {
        _aggregateId = Guid.NewGuid();
        _store = DocumentStore.For(options =>
        {
            options.Connection("Server=localhost;Database=multi-tenancy-issue;Port=5432;User Id=postgres;Password=postgres;");

            options.Events.TenancyStyle = TenancyStyle.Conjoined;
            options.Events.AddEventType<CreatedEvent>();
            options.Events.AddEventType<UpdatedEvent>();

            options.Projections.SelfAggregate<MyAggregate>(ProjectionLifecycle.Inline).MultiTenanted();
        });
        
        // Create a new aggregate in the default tenant
        await using (var session = _store.OpenSession())
        {
            var createdEvent = new CreatedEvent("Created");

            session.Events.StartStream<MyAggregate>(_aggregateId, createdEvent);
            await session.SaveChangesAsync();
        }
    }

    [Test]
    public async Task Appending_Events_To_Default_Tenant_From_Other_Tenant_Should_Succeed()
    {
        // Open a session for another tenant
        await using var session = _store.OpenSession("MyTenant");

        // Append new event in the default tenant
        session.ForTenant(Tenancy.DefaultTenantId).Events.Append(_aggregateId, new UpdatedEvent("Updated"));
        await session.SaveChangesAsync(); // SaveChanges loads the projected document from the wrong tenant when applying the new events
        
        // Load the projected document from the default tenant
        var aggregate = await session.ForTenant(Tenancy.DefaultTenantId).LoadAsync<MyAggregate>(_aggregateId);
        
        // Both the Created and Updated fields should be filled in
        Assert.NotNull(aggregate?.Updated, "aggregate.Updated is null"); // Updated is filled in because the UpdatedEvent is applied correctly
        Assert.NotNull(aggregate?.Created, "aggregate.Created is null"); // Created is null because the document was loaded from the wrong tenant when applying the Updated event
    }
}

```

Examining the Marten logs also show that while the events are appended and saved to the correct tenant, the document is loaded from the wrong tenant.

```
select d.id, d.data from public.mt_doc_myaggregate as d where id = :id and d.tenant_id = :tenantid
  id: 88411742-79e3-4c7b-8a52-e6e2d76db8f3
  tenantid: MyTenant //Loaded from wrong tenant

select public.mt_upsert_myaggregate(:p0, :p1, :p2, :p3, :p4);update public.mt_streams set version = :p5 where id = :p6 and version = :p7 and tenant_id = :p8;insert into public.mt_events (data, type, mt_dotnet_type, seq_id, id, stream_id, version, timestamp, tenant_id) values (:p9, :p10, :p11, :p12, :p13, :p14,
:p15, :p16, :p17);
  p0: {"Id":"88411742-79e3-4c7b-8a52-e6e2d76db8f3","Created":null,"Updated":"Updated"}
  p1: MartenMultitenancyBug.MyAggregate
  p2: 88411742-79e3-4c7b-8a52-e6e2d76db8f3
  p3: 0183a397-7717-46e5-88df-042cb7694db5
  p4: *DEFAULT* // Saved in correct tenant
  p5: 2
  p6: 88411742-79e3-4c7b-8a52-e6e2d76db8f3
  p7: 1
  p8: *DEFAULT*
  p9: {"NewName":"Updated"}
  p10: updated_event
  p11: MartenMultitenancyBug.UpdatedEvent, MartenMultitenancyBug
  p12: 14
  p13: 0183a397-7711-4e0a-aa81-e8f1e761ef23
  p14: 88411742-79e3-4c7b-8a52-e6e2d76db8f3
  p15: 2
  p16: 04/10/2022 15:22:36 +00:00
  p17: *DEFAULT*

```

**Comment by jeremydmiller** [comment]
Duplicate cause with #2353. Just a bandwidth issue. I know the root cause, but the fix was probably going to be involved. I'll get to it soon.

**Comment by jeremydmiller** [comment]
Being worked by @oskardudycz. There's a PR in progress to address this

**Comment by WilvanBil** [comment]
@oskardudycz So when will 6.0.0 be available? I have a workaround for this issue, but it would be lovely if it was fixed in a release.

**Comment by oskardudycz** [comment]
It's already available in `v6.0.0-rc.1`, the final release should be made in a week or two.

**Comment by mysticmind** [comment]
The issue outlined itself is a bit of a convoluted/complicated use-case and I am able to reproduce this issue in the latest Marten 6.0.3.

The following are my observations:
 - Session uses tenant`MyTenant`  via the vall `await using var session = _store.OpenSession("MyTenant");`
 - `session.ForTenant(Tenancy.DefaultTenantId).Events.Append(_aggregateId, new UpdatedEvent("Updated"));` appends adds an event using default tenant.
 - `options.Projections.SelfAggregate<MyAggregate>(ProjectionLifecycle.Inline).MultiTenanted();` or using the new equivalent `options.Projections.Snapshot<MyAggregate>(SnapshotLifecycle.Inline).MultiTenanted();`. Updating of snapshot attempts to load the document using tenant `MyTenant` and then apply the events. Since the aggregate document is not present in `MyTenant`, it instantiates a new object and applies only the last event.
 
 In summary, loading of the snapshot document will need to use the content of tenant under which the event was added.

Re-opening the same.

**Comment by haefele** [comment]
Hey, I'm also running into this issue using Marten 6.2.0.
The problem only occurs to me when I'm using [Custom Aggregation Projections](https://martendb.io/events/projections/custom-aggregates.html) and not for [Single Stream Projections](https://martendb.io/events/projections/aggregate-projections.html).

I was digging through the code a bit, and it seems like the reason Single Stream Projections use the correct tenant for operations is [this](https://github.com/JasperFx/marten/blob/21e172af3915123b1274926de98b516223b506bd/src/Marten/Events/Aggregation/AggregationRuntime.cs#L58) call to `UseTenancyBasedOnSliceAndStorage`.

For Custom Aggregation Projections it seems like it doesn't do that [here](https://github.com/JasperFx/marten/blob/21e172af3915123b1274926de98b516223b506bd/src/Marten/Events/Aggregation/CustomProjection.cs#L48).

Is this already the problem? Would you accept a PR for this?

**Comment by jeremydmiller** [comment]
@haefele You've hit a permutation nobody has considered before. If you're up for a PR, yes, please

**Comment by jeremydmiller** [comment]
Looks like this was fixed already by https://github.com/JasperFx/marten/pull/2764

**Comment by iamjustsam** [comment]
Hey, it looks like this issue still persists in Marten 7.6.0 using the same test case as above.

**Comment by haefele** [comment]
Hey, that's a bit weird, I thought I fixed this issue in #2764. 
In that PR I also added a test to marten itself called `ForEventsAppendedToTenantedSession_CustomProjection`.
Maybe you can compare your own code to that test to see what's going on?

As far as I can tell, the test still exists and completes successfully.
